### DISCUSSION TITLE
Idea - advanced shader script field for making pixi filters and sharing them as extensions

### DISCUSSION THREAD

**Discussion by blurymind** [4ian/GDevelop#2340] [Enhancement ideas (and their technical discussions)]
So this is kind of a stretch, but what if we could write our pixi filters straingt inside gdevelop? 
We already have monaco for editing, all we need then is a game object that can read the vertex shader code and render it.

There are already some open source apps that could let us edit shaders in real time and see the result
https://github.com/gkjohnson/webgl-shader-editor
try it here 
https://gkjohnson.github.io/webgl-shader-editor/dist/index.bundle.html
Imagine this with one of your game sprites/animations in the previewer window

here is a video of how to set up a shader for pixi:
https://www.youtube.com/watch?v=wIC-CqsUplw

So now imagine the filter editor window having an extra option to add a custom shader and the ability to edit it via an external shader editor (similar to piskel's bundling) or monaco itself

This could lead to more shaders being made for gdevelop :) not just the pixi collection :)
Even pixi shaders made with gdevelop

**Comment by 4ian** [comment]
Basically, much like you can create actions/conditions/expressions/behaviors in an extension directly from GDevelop, there could be a way to do it for effects.
That could be interesting indeed - even if there won't be visual events to do that. Though some editors like Unity do have something interesting called "shader graphs" (maybe actually events could be used in the future to do shaders?? :)).

**Comment by blurymind** [comment]
Using the event sheet to create shaders.. interesting :) 
This is something that gdevelop can potentially be the first one to do. One problem we currently have with filters is that the IDE is not letting us preview them in any way - one has to run the game to see the result.

Other engines update the shader in the viewport as you edit it in real time. Godot is quite good at this.

In terms of visual programming, shaders have always used a node style connection approach.

I would personally be quite pleased if as a starting point we can at least edit shader code in the IDE and see it update without having to re-run the game. If it's done in a pixi friendly way I think we might attract more pixi devs to gdevelop - to use it as a tool to r&d pixi filters

**Comment by Quarkstar** [comment]
This is really cool:) 
Creating filters is much easier in PixiJs v5. (May I ask when we willI upgrade to pixi v5?)

It seems editing shader code is not hard to implement.
https://github.com/pixijs/pixi.js/wiki/v5-Creating-filters
```javascript
new Filter(vertShader, fragShader, myUniforms); //myUniforms is a json
```

And here is an example.
https://pixijs.io/examples/#/filters-advanced/custom.js

**Comment by blurymind** [comment]
There is currently a task on the Trello board that nobody has picked yet.

If you upgrade pixi, it will help me also with my tilemap work, since later
on I don't have to update pixi-tilemap to the p5 version

**Comment by Quarkstar** [comment]
Ok, I would like to give it a try. 
There seems to be many testings to do and it may take some time :joy:.
I will do it as fast as I can.

**Comment by blurymind** [comment]
We are already upgrading electron, might as well bump up pixi :)

**Comment by 4ian** [comment]
I made an issue for Pixi v5 update here: https://github.com/4ian/GDevelop/issues/1605

**Comment by Bouh** [comment]
For curious i use shader with filters in my dev branch for make a color transparent like we can do on GD4.
[Here the sandbox ](https://jsfiddle.net/Babar31/vqyk8gmf/4/)with my shader for make a color transparent in pixi v4.8.6.

[Here my branch,] it's works on GD5, you need compile libGD.js before.(https://github.com/Bouh/GDevelop/tree/rewrite_method_in_RuntimeSpriteObject_for_GD5)

**Comment by Bouh** [comment]
My vision for effect/shader, see that like the extension but it's custom filter for layer and object!

Here two mockups, one for code editor in full tab.
Second the most interesting, the effect editor.

![idea_code_editor_GDevelop](https://user-images.githubusercontent.com/1670670/78087528-a6a5fd80-73c1-11ea-823d-e8a5f85113f2.png)

![idea_effect_editor_GDevelop](https://user-images.githubusercontent.com/1670670/78088106-7cedd600-73c3-11ea-9ba8-281cbe8a1e89.png)



On tabs you can see the new types.
Possibility to open files present as ressource directly from resources tab in to a code editor.

On second mockup you see list of effect, and how configure the effect with name and uniforms.
About uniforms we can provide to shader one or multiple image(s), so naturally i think see the type texture and select a texture with the resource selector from GD!

**Comment by blurymind** [comment]
It would've great if like Godot we can see the filters change as we edit
their shader code.

**Comment by 4ian** [comment]
> My vision for effect/shader, see that like the extension but it's custom filter for layer and object

Yeah that would be a nice addition to have - can you create a card on the roadmap?

> It would've great if like Godot we can see the filters change as we edit
their shader code.

For this problem of "I want to shorten the feedback loop between doing changes to my effects and seeing the result" I'm thinking of three things:
1. Displaying them in the scene editor
2. and/or displaying a preview like what Bouh is suggesting (could work both for the "effect editor" that Bouh is describing, and also when editing effects in the list of effects of a layer or an object).
3. and/or have effects hot reloaded in a preview :)


Solutions 1. and 2. would need some careful design to be able to use the same JS files to both have an effect in the game engine (what we're doing now) and also be able to use the effect in the editor (i.e: make an equivalent of `gdjs.PixiFiltersTools.registerFilterCreator` that work in the editor).

Solution 3. is a particularly interesting idea that goes beyond effects. We could imagine tweaking the game engine and editor so that if the editor regenerate the preview, the game engine gets a signal and apply the changes in events, effects, etc... without restarting the game.
It's a card that I've added here: https://trello.com/c/Ce46QxKj/417-add-support-for-hot-reload-when-a-preview-is-running-and-play-button-is-clicked-continue-running-the-game-with-updated-events
It's a longer term stuff, but basically we make the preview the best way to see your changes in real time :)

**Comment by blurymind** [comment]
I would much rather be able to preview the filter's changes in the Ide somehow. The hot reloading solution sounds like a bigger task that would give us other gains though :) 

In terms of UIX, with @Bouh 's mockups- I dont like switching between tabs when working on a shader. I want the fragment code, the vertex code and the export parameters to all be in the same tab with where I can preview my changes. I dont want to have a situation where I am constantly flipping between tabs to see what's happening. Dont separate them into tabs please. They are all one thing.
Again godot is the best example here - you have everything in one window.You change something, you get instant feedback and you dont need to switch a single tab. Your code is typed below the scene view. Your parameters are on the right side inspector. Not a single click required to display one or the other. Everything happens under one roof

**Comment by Bouh** [comment]
>I dont like switching between tabs when working on a shader

It's a good reason i agree!
So here the new idea, with a new little icons in tabs for extend to another full tabs. This will be cool in future with modal window from electron! Tho this is mostly important for JSEvent. 

![image](https://user-images.githubusercontent.com/1670670/78252559-5464f980-74f3-11ea-85a4-6fcb4579eb12.png)

![image](https://user-images.githubusercontent.com/1670670/78566367-66120e00-781f-11ea-9f68-b47bb1655ce8.png)

**Comment by blurymind** [comment]
> > I dont like switching between tabs when working on a shader
> 
> It's a good reason i agree!
> So here the new idea, with a new little icons in tabs for extend to another full tabs. This will be cool in future with modal window from electron! Tho this is mostly important for JSEvent.
> 
> ![image](https://user-images.githubusercontent.com/1670670/78252559-5464f980-74f3-11ea-85a4-6fcb4579eb12.png)

Yes I like that layout very much :) 

In general it would be cool if we could detach any tab as a new electron window - but that should be a separate task imo.

Some people have complained that piskel and yarn are in modal windows,as they expressed desire to have their editor in their second monitor and be able to work in the IDE and the external editor without locking the IDE.

It would be nice if at some point we remove their modality. But again- separate task imo

**Comment by 4ian** [comment]
Agree that I prefer everything related to a concept (here an Effect, aka filter, aka "shader") in a single tab :) 
(We'll maybe be able at some point to separate tabs in separate windows (I did some experiments, which kind of worked for electron and the web-app, but this needed a bit of hack to get CSS styles to propagate between dialogs))
